### ISSUE TITLE
nans while using chair dataset

### ISSUE THREAD

**Issue by Tetragramm** [JonathanRaiman/infogan#7]
Exactly as the title says, I am getting NaN values pretty quickly.

The MNIST test works pretty well, though not perfectly.  There's some swapping between numbers in the categories.  I don't know if that's related.

On the chairs dataset, I am using the command from the README.  I don't know what circumstances cause the NaNs, it seems to happen at a random time.  The first attempt failed at epoch about 80, the second at epic 1, and my last attempt was epoch 6.  

I think it's a fundamental problem though.  I managed to get one run through to epoch 80, which I had set as the total number, but the output images it produced were just nonsense.  Like a really bad deconvolution.  Unfortunately while I was debugging I reverted all the changes, including the log folder, so I can't show you that.

I'm not actually sure where to begin trying to debug this, but I'm happy to help if there's anything you need.  I'm running the latest version of tensorflow-gpu, the latest NVIDIA drivers, Windows 10.  If there's anything I can do to help, just ask.

**Comment by JonathanRaiman** [comment]
@Tetragramm Thanks for sharing this information about running on the chairs dataset. I suspect NaNs are occurring due to some imbalance in generator/discriminator training; Have you tried different learning rates to find one where the loss for either component does not hit an inf/nan?

**Comment by Tetragramm** [comment]
I have not tried that. I have been updating drivers and so forth to make sure there's not problems there.

The rates are still at the paper recommended generator 1.0e-3 and discriminator 2.0e-4.  I would think lower rates are less likely to Nan, but do I change just the discriminator (which NaNs first) or do I change both to keep the proportions about the same?  The paper is vague on how the rates were determined in the first place (as always) so I'm not sure how important the relative proportions are.  

Question:  The plot_every command is how many images to process before generating the images in the log.  Is it possible to have it not overwrite the old ones so I get a sense of how it's progressing from the images?  I don't see any way of controlling that.

**Comment by JonathanRaiman** [comment]
I'm not sure about the images being overwritten -- usually the tf.image_summary takes incrementing iteration numbers as new versions and should place them side by side. You can also save the images to disk (adding a plt.savefig somewhere in `categorical_grid_plots.py`).

For the learning rates, to be on the safe side reduce both as low as you can so that some progress is still being made (e.g. higher than 1e-9 or something). Higher learning rates means faster training, but as a counterpart, higher risk of instability/jumping out of the current parameter space/NaNs. Thus you should try to grow them from the smallest point where no "breaking" happens until the convergence/training is achieved in reasonable time. The ratio of the learning rates can be determined as follows: you want the losses to be on par for both models (e.g. in tboard). If the generator's loss is high, then discriminator lr is too high, or gen discriminator is too low and vice-versa. (usually discriminator has an easier job, so the discriminator's learning rate is usually set to a lower value as a "handicap").

**Comment by Tetragramm** [comment]
All right, I just completed a 100 epoch run with the default learning rates with no NaNs, but I still got bad results. I've linked screenshots of the graphs and one of the resulting images.  All of the others from that iteration look the same (and I don't see any way of viewing other iterations).
[http://imgur.com/a/lmCiS](http://imgur.com/a/lmCiS)

Tonight I'll start a lower rate learning model and see how that works.

**Comment by Tetragramm** [comment]
All right, I've been decreasing it by an order of magnitude each run, and I just finished the 1.0e-6 and 2.0e-7 run.  I'm still getting NaNs.

Any ideas?

**Comment by JonathanRaiman** [comment]
Hm. No strong guesses at this point. I'm not sure I will have time this weekend to take a look. My process would be as follows:
Typically two events can cause this behavior:
1. too big steps (solution: lower lr)
2. unstable or explosive training [e.g. values near 0 sent to a log or a division] (solution: use grad clipping or epsilons inside log/division).

Since you've tried case 1, case 2 seems likely. I would suggest monitoring the gradient norm during training, (print it before doing `apply_gradients` or send it to tensorboard) add 1e-6 or 1e-9 to the logit before inputting to the log-loss. If gradients increase dramatically before NaNs, then this is the issue, and efforts should be taken to make sure the model is either better able to train from the get-go, or find ways or skipping/preventing NaN values from occurring through the use of epsilons inside nan-causing functions.
